### ISSUE TITLE
TDMML: Tacrolimus

### ISSUE THREAD

**Issue by aegis301** [aidh-ms/clinical-study-proposals#2]
# Predicting Tacrolimus Plasma Levels Using Machine Learning

## Principal Investigator(s)
AIM-Group

## Study Type
Retrospective observational study

---

## Background & Clinical Significance

Tacrolimus is a critical immunosuppressant administered to patients after major organ transplantation to prevent rejection. It requires careful individualized dosing due to its narrow therapeutic window - inadequate levels risk organ rejection while excessive levels cause toxicity (nephrotoxicity, neurotoxicity, metabolic complications). 

ICU patients present additional challenges: rapidly changing physiology, organ dysfunction, drug-drug interactions, and variable absorption make achieving target levels difficult. Current practice relies on clinical judgment, often resulting in out-of-range levels that require frequent dose adjustments.

### Current Clinical Workflow

**Daily Decision Timeline:**
- **06:00** - Blood drawn for tacrolimus trough level measurement; morning dose administered
- **13:00** - Interdisciplinary team (surgeons + intensivists) reviews patient status and morning labs, decides evening dose
- **20:00** - Evening dose administered
- **Next day 06:00** - Blood drawn to measure resulting tacrolimus level

The critical decision point is at **13:00** when clinicians must predict what evening dose will achieve target levels the next morning, based on current patient status, recent labs, and dosing history.

### Proposed Solution

We aim to develop a machine learning model that predicts next-morning tacrolimus levels based on patient status and proposed evening dose. This will be implemented as an **interactive dashboard** where clinicians can "test" different evening doses and see predicted outcomes, enabling data-driven dose optimization.

---

## Primary Research Question

Can we develop a machine learning model to accurately predict next-morning tacrolimus trough levels (RMSE < 2.0 ng/mL) in ICU patients, enabling clinicians to optimize evening dose selection through an interactive decision support tool?

---

### Inclusion Criteria
- Adult patients (≥18 years) admitted to ICU
- Receiving tacrolimus via oral (p.o.) or sublingual route
- At least one tacrolimus trough level measurement
- At least 24 hours of ICU stay prior to measurement

### Exclusion Criteria
- Tacrolimus administered via continuous infusion (different pharmacokinetics)
- Missing critical baseline data (weight, height)
- Age < 18 years

### Expected Sample Size
- Approximately 8,000 patients
- Approximately 44,000 tacrolimus level measurements
- Average 5.5 measurements per patient

---

## Primary Outcome Variable

**Tacrolimus trough level** (continuous variable, ng/mL) measured at 06:00 on day t+1.

**Therapeutic range:** 5-15 ng/mL (may vary by organ type and time post-transplant)

**Note on administration routes:** 
- Oral (p.o.) doses recorded as-is
- Sublingual doses multiplied by 3 to account for bioavailability differences (based on institutional practice)

---

### Temporal Framework

**Decision Point:** Day t, 13:00 (1 PM)  
**Feature Window:** Day t, 00:00-12:00 (midnight to noon)  
**Target:** Day t+1, 06:00 tacrolimus level

All features represent data available at the decision point (day t, 13:00).

### 1. Patient Demographics (Baseline)
- Age (years, at ICU admission)
- Sex (categorical: M/F)
- Weight (kg)
- Height (cm)
- BMI (calculated)
- Days since ICU admission

### 2. Comorbidities (from medical history)
- Diabetes mellitus (boolean)
- Chronic kidney disease (boolean)
- Heart failure (boolean)
- Transplant type (categorical: kidney/liver/heart/other - from OPS codes if available)

### 3. Laboratory Values (Day t, 00:00-12:00, mean if multiple)
- **Renal function:** GFR (mL/min), Creatinine (mg/dL)
- **Hepatic function:** Bilirubin (mg/dL), GOT/AST (U/L), GPT/ALT (U/L), Albumin (g/dL)
- **Hematology:** Hematocrit (%), Hemoglobin (g/dL)
- **Inflammation:** CRP (mg/L), Lactate (mmol/L)
- **Electrolytes:** Magnesium (mmol/L) - affects tacrolimus
- **Other:** INR (if available)

### 4. Clinical Status (Day t, 00:00-12:00)
- **Vasopressor support:** Vasopressor requirement in the last 12h (boolean)
- **Respiratory support:** Mechanical ventilation (boolean)

### 5. Tacrolimus Dosing History (Critical Features)

**Recent doses (adjusted for route):**
- Day t morning dose (already given, mg)
- Day t-1 evening dose (mg)
- Day t-1 morning dose (mg)
- Day t-2 evening dose (mg)
- Day t-2 morning dose (mg)
- Day t-3 evening dose (mg)
- Day t-3 morning dose (mg)

**Recent levels:**
- Day t-1 trough level (ng/mL)
- Day t-2 trough level (ng/mL)
- Day t-3 trough level (ng/mL)

**Dose stability indicators:**
- Coefficient of variation in last 3 days
- Number of dose changes in last 5 days

### 6. Drug-Drug Interactions (Day t, cumulative dose 00:00-12:00)

**Other relevant medications:**
- Proton pump inhibitors (might affect absorption) - boolean
- Metamizole - cumulative dose in g

### 7. Key Input Variable (To Be Optimized)
**Day t evening dose** (mg) - This is the decision variable that clinicians will adjust in the dashboard

---

### Workflow at Decision Point (Day t, 13:00)

1. **Auto-populate known data:**
   - System retrieves patient demographics, morning labs, recent doses/levels
   - Displays current tacrolimus trend (graph of last 7 days)

2. **Interactive dose optimization:**
   - Clinician enters proposed evening dose (e.g., 3 mg)
   - Model predicts: "Expected level tomorrow: 11.2 ng/mL (within range ✓)"
   - Clinician adjusts dose (e.g., tries 4 mg)
   - Model updates: "Expected level tomorrow: 14.8 ng/mL (upper limit)"
   - Clinician selects optimal dose

3. **Decision support features:**
   - Confidence intervals around prediction
   - Risk alerts (e.g., "High risk of supratherapeutic level")
   - Comparison to previous day's prediction vs. actual

4. **Feedback loop:**
   - Next morning: actual level measured
   - System calculates prediction error
   - Model continuously improves with new data

---

### Data Preprocessing

1. **Route adjustment:** Multiply sublingual doses by 3
2. **Missing data handling:** 
   - Impute lab values using forward-fill (last observation carried forward)
   - Create missingness indicators for key variables
   - Exclude measurements with >30% missing features
3. **Outlier detection:** Remove physiologically implausible values
4. **Feature scaling:** Standardize continuous variables (mean=0, SD=1)

### Handling Multiple Measurements Per Patient

**Critical methodological consideration:** Our dataset contains ~44,000 measurements from ~8,000 patients (average 5.5 per patient). These are not independent observations.

**Problem:** Standard random train-test split would place the same patient in both sets, causing:
- Data leakage (model learns patient-specific patterns)
- Overly optimistic performance estimates
- Poor generalization to new patients

**Solution: Patient-Level Splitting**
	see comment below
### Model Development Strategy

**Phase 1: Baseline Models**
1. **Linear regression with regularization** (Lasso/Ridge)
   - Interpretable baseline
   - Identifies most important features
   
2. **Pharmacokinetic model** (one-compartment)
   - Domain knowledge baseline
   - Incorporates half-life (~12 hours)

**Phase 2: Advanced Models**
3. **Random Forest**
   - Handles non-linearity
   - Provides feature importance
   - Robust to outliers

4. **Gradient Boosting** (XGBoost, LightGBM)
   - Often best performance
   - Handles complex interactions
   - Fast training/prediction

5. **Neural Networks** (if sufficient data)
   - Deep learning approach
   - Captures complex patterns

**Phase 3: Ensemble & Calibration**
6. **Stacked ensemble** of best models
7. **Bayesian approaches** for uncertainty quantification
8. **Conformal prediction** for prediction intervals

### Model Evaluation Metrics

**Primary Metrics:**
- **RMSE** (Root Mean Square Error) - penalizes large errors
- **MAE** (Mean Absolute Error) - average prediction error
- **R²** (coefficient of determination) - proportion of variance explained

**Secondary Metrics:**
- **MAPE** (Mean Absolute Percentage Error)
- **Within ±2 ng/mL accuracy** (clinical threshold)
- **Within therapeutic range accuracy** (5-15 ng/mL)

**Stratified Performance Analysis:**
- By transplant type (kidney vs. liver vs. heart)
- By renal function (GFR categories)
- By dose stability (stable vs. fluctuating)
- By time since admission (early vs. late ICU stay)

**Clinical Utility Metrics:**
- Reduction in out-of-range measurements
- Reduction in dose adjustments needed
- Time to achieve stable therapeutic levels

### Feature Importance Analysis

- SHAP (SHapley Additive exPlanations) values
- Permutation importance
- Partial dependence plots
- Identify key drivers of tacrolimus levels

---

### Source Tables (PostgreSQL Database)

1. **events** - Lab values, vital signs, scores
2. **medications** - Tacrolimus doses (hourly, aggregate to daily)
3. **icu_stay** - Demographics, admission/discharge dates
4. **medical_history** - Comorbidities
5. **ops** - Procedure codes (identify transplant type)

### Key Item IDs to Identify
- Tacrolimus level measurement (events.itemid)
- Tacrolimus medication (medications.itemid)
- Route of administration (medications.access: 'oral' vs 'sublingual')
- All lab values (GFR, bilirubin, etc.)
- Drug-drug interaction medications

### SQL Query Structure
```sql
-- Pseudocode outline:
1. Identify all tacrolimus measurements (trough levels at 06:00)
2. For each measurement, extract:
   - Patient demographics from icu_stay
   - Lab values from events (day t, 00:00-12:00, mean)
   - Tacrolimus doses from medications (t, t-1, t-2, t-3)
   - Previous tacrolimus levels (t-1, t-2, t-3)
   - Interacting medications (day t cumulative)
3. Apply route adjustment (sublingual × 3)
4. Create lag features using window functions
5. Filter for inclusion/exclusion criteria
```

**Comment by aegis301** [comment]
## Handling Multiple Measurements Per Patient

### The Challenge
Our dataset contains approximately 44,000 tacrolimus measurements from 8,000 patients (average ~5.5 measurements per patient). These measurements are not independent observations - measurements from the same patient are more correlated with each other than with measurements from different patients.

### The Problem with Naive Splitting
If we perform a standard random train-test split across all 44,000 measurements, we risk:

1. **Data leakage**: The same patient appears in both training and test sets
2. **Overly optimistic performance**: The model learns patient-specific patterns rather than generalizable relationships
3. **Poor real-world performance**: The model performs well on known patients but fails on new patients

**Example**: If Patient X has 10 measurements and we randomly split them, 8 might end up in training and 2 in testing. The model learns "Patient X typically has levels around 10 ng/mL" and appears to predict well on the test set - but this doesn't reflect how it will perform on a completely new patient.

### Our Approach: Patient-Level Splitting

We will split data at the **patient level**, not the measurement level:

1. **Training set**: All measurements from 80% of patients (~6,400 patients, ~35,000 measurements)
2. **Test set**: All measurements from 20% of patients (~1,600 patients, ~9,000 measurements)
3. **No patient appears in both sets**

This ensures the model must generalize to completely new patients, providing realistic performance estimates for clinical deployment.

### Cross-Validation Strategy

We will use **GroupKFold cross-validation** with patient IDs as groups:
- 5-fold cross-validation
- Each fold contains different patients
- No patient appears in both training and validation within any fold
- This accounts for within-patient correlation in our performance estimates

### Additional Consideration: Temporal Validation

If feasible, we will also perform a temporal validation split:
- Training: All measurements before a specific date
- Testing: All measurements after that date

This simulates real-world deployment where the model is trained on historical data and must predict for future patients.

### Statistical Implications

By using patient-level splitting and grouped cross-validation, we:
- Obtain conservative, realistic performance estimates
- Ensure the model generalizes to new patients
- Avoid inflated metrics due to within-patient correlation
- Prepare the model for actual clinical deployment
